helix-core/src/graphemes.rs

```diff
@@ -5,7 +5,79 @@ use ropey::{iter::Chunks, str_utils::byte_to_char_idx, RopeSlice};
 use unicode_segmentation::{GraphemeCursor, GraphemeIncomplete};
 use unicode_width::UnicodeWidthStr;
 
-use std::fmt;
+use std::borrow::Cow;
+use std::fmt::{self, Display};
+
+use crate::LineEnding;
+
+#[derive(Debug, Clone, PartialEq, Eq)]
+/// A preprossed Grapheme that is ready for rendering
+pub enum Grapheme<'a> {
+    Space,
+    Newline,
+    Nbsp,
+    Tab { width: u16 },
+    Other { raw: Cow<'a, str>, width: u16 },
+}
+
+impl<'a> Grapheme<'a> {
+    pub fn new(raw: Cow<'a, str>, visual_x: usize, tab_width: u16) -> Grapheme<'a> {
+        match &*raw {
+            "\t" => {
+                let width = tab_width - (visual_x % tab_width as usize) as u16;
+                Grapheme::Tab { width }
+            }
+            " " => Grapheme::Space,
+            "\u{00A0}" => Grapheme::Nbsp,
```

Comment:
So.. I mean if we want to pay attention to nbsp, how correct do we want to be here? Because nbsp is not the only place where line breaks are forbidden with a Unicode grapheme. There is 202F NARROW NO-BREAK SPACE (NNBSP), U+2060 WORD JOINER (WJ), 180E MONGOLIAN VOWEL SEPARATOR (MVS), and many others. The [Unicode line breaking algorithm](https://www.unicode.org/reports/tr14/#Algorithm) is quite complex. I know we probably don't want to go to that extent, but what is the primary use case for looking at just one Unicode character with a non-breaking property?

Reply:
I experimented a bit with unicode-linebreaks but the API is too inflexible (no streaming API and the unicode line wrapping algorithm is also not truly linear and very hard to apply here in general. Some rules also seem a bad fit for programming languages (breaks are prohibited at brackets).

I think future work could focus on a better algorithm but for now I am simply using `!char_is_word` to detect word breaks and use that as separators.